client.go

```diff
@@ -69,28 +71,155 @@ type Client struct {
 	// through legitimate channels.
 	dopplegangerAddrs map[string]struct{}
 	badPeerIPs        map[string]struct{}
-	torrents          map[InfoHash]*Torrent
+	torrents          map[InfoHash]*torrent
 
 	acceptLimiter   map[ipStr]int
 	dialRateLimiter *rate.Limiter
 }
 
 type ipStr string
 
+// BadPeerIPs ...
 func (cl *Client) BadPeerIPs() []string {
 	cl.rLock()
 	defer cl.rUnlock()
 	return cl.badPeerIPsLocked()
 }
 
+// Start the specified torrent.
+// Start adds starts up the torrent within the client downloading the missing pieces
+// as needed. if you want to wait until the torrent is completed use Download.
+func (cl *Client) Start(t Torrent) (dl Download, added bool, err error) {
+	return cl.start(t)
+}
+
+// MaybeStart is a convience method that consumes the return types of the torrent
+// creation methods: New, NewFromFile, etc. it is all respects identical to the Start
+// method.
+func (cl *Client) MaybeStart(t Torrent, failed error) (dl Download, added bool, err error) {
+	if failed != nil {
+		return dl, false, failed
+	}
+	return cl.start(t)
+}
+
+// Download the specified torrent.
+// Download differs from Start in that it blocks until the torrent is completely downloaded.
+// the provided ctx is used to timeout the download, but it currently only works with regards
+// to fetching the torrent info.
+func (cl *Client) Download(ctx context.Context, t Torrent) (dl Download, added bool, err error) {
+	var (
+		dlt *torrent
+	)
+
+	if dlt, added, err = cl.start(t); err != nil {
+		return dlt, added, err
+	}
+
+	select {
+	case <-dlt.GotInfo():
+	case <-ctx.Done():
+		return nil, added, ctx.Err()
+	}
+
+	dlt.DownloadAll()
+
+	// cheat here by leveraging the reader to block while copying
+	if _, err := io.Copy(ioutil.Discard, dlt.NewReader()); err != nil {
+		return dlt, added, err
+	}
+
+	return dlt, added, nil
+}
+
+func (cl *Client) start(t Torrent) (dlt *torrent, added bool, err error) {
+	if dlt = cl.download(t); dlt != nil {
+		l2.Println("torrent found returning")
+		return dlt, false, cl.merge(dlt, t)
+	}
+
+	cl.lock()
+	dlt = cl.newTorrent(t)
+	cl.unlock()
+
+	cl.merge(dlt, t)
+
+	cl.lock()
+	defer cl.unlock()
+	cl.eachDhtServer(func(s *dht.Server) {
+		go dlt.dhtAnnouncer(s)
+	})
+	cl.torrents[t.InfoHash] = dlt
+	cl.clearAcceptLimits()
+	dlt.updateWantPeersEvent()
+	// Tickle Client.waitAccept, new torrent may want conns.
+	cl.event.Broadcast()
+
+	return dlt, true, nil
+}
+
+// Stop the specified torrent, this halts all network activity around the torrent
+// for this client.
+func (cl *Client) Stop(t Torrent) (err error) {
+	cl.lock()
+	defer cl.unlock()
+	return cl.dropTorrent(t.InfoHash)
+}
+
+// SetMaxEstablishedConns for the torrent.
+func (cl *Client) SetMaxEstablishedConns(t Torrent, max int) (oldMax int) {
```

Comment:
this is a minor artifact where I was trying to reduce the number of methods publicly exposed on the original Torrent struct.

leaning towards changing this method to something along the lines of:
```golang
func (cl *Client) Tune(t Torrent, options ...Tuning) error {
  // applying tuning configuration to the specified torrent.
}
```